### DISCUSSION TITLE
Does this library have an event that triggers on node deletion?

### DISCUSSION THREAD

**Discussion by DoaaRadwan** [bcakmakoglu/vue-flow#1176] [Q&A]
Hi,

Is there an event that triggers on node deletion, pressing the `backspace` and whatever the delete key is set to? like `onNodeDelete` or something similar.

We need to implement additional logic on deletion, so currently we have the delete key disabled and rely on a custom delete button on our modals but it would be really helpful to the users to see the same behaviour of deletion whether they press the keyboard key or click the custom delete button.

**Comment by bcakmakoglu** [comment]
There is an event for changes, called `onNodesChange` which will receive an array of changes as an arg.

```ts
const { onNodesChange } = useVueFlow()

onNodesChange(changes => {
	changes.forEach(change => {
		console.log(change.type) // 'position' | 'add' | 'remove' | 'select' | 'dimensions'
	})
})

**Reply by DoaaRadwan** [reply]
`change` will only return the id of the node after the node has been deleted - so there is no way to access other info contained in the node object, I would like to access the node properties. Is there a way for `onNodesChange` to return the node object ?

Thanks for the quick answer!

**Reply by DoaaRadwan** [reply]
We would like to display a confirm modal for the user before the deletion

**Reply by bcakmakoglu** [reply]
This happens because changes are, by default, *auto-applied*.
You can disable this behavior and instead manually apply changes, giving you full control over when and why to apply them.

```vue
const { onNodesChange, onEdgesChange, applyNodeChanges, applyEdgeChanges } = useVueFlow()

onNodesChange(changes => {
	// evaluate your changes

	applyNodeChanges(evaluatedChanges)
})

// same for onEdgesChange

<VueFlow v-model="elements" :apply-default="false">...</VueFlow>
```

**Reply by DoaaRadwan** [reply]
I have added `:apply-default="false"` to VueFlow tag

and have this function 
```
onNodesChange((changes) => {
  changes.forEach((change) => {
    if (change.type == "remove") {
      openDeleteModal();
    } else {
      applyNodeChanges([change]);
    }
  });
});
``` 
but the deletion seems to happen regardless - before the modal opens

**Reply by bcakmakoglu** [reply]
[Works fine for me in a sandbox using a simple alert.](https://codesandbox.io/p/sandbox/runtime-dew-tx34mg?file=%2Fsrc%2FApp.vue%3A15%2C37)

**Reply by DoaaRadwan** [reply]
Mmm.. I have been trying since yesterday with no luck - we are using something similar to the drag and drop example on vue js (no typescript) and the deletion still happens even when `applyNodeChanges([change]);` is not used at all

**Reply by bcakmakoglu** [reply]
Are you using multiple ways to pass options, so `applyDefault` might be overwritten somehow?

**Reply by DoaaRadwan** [reply]
What counts as passing options ?(please excuse my inexperience). 
I have this 
```
<VueFlow
      :class="{ dark }"
      :apply-default="false"
      :default-zoom="1"
      :min-zoom="0.2"
      :max-zoom="4"
      class="basicflow"
      fit-view-on-init
      @dragover="onDragOver"
    >
```
and this 
```
const {
  onNodeDragStop,
  onConnect,
  addEdges,
  addNodes,
  onEdgeDoubleClick,
  onNodeDoubleClick,
  removeEdges,
  removeNodes,
  findNode,
  project,
  vueFlowRef,
  nodes,
  edges,
} = useVueFlow(elements);
```
where `elements` is an object of `{nodes: [], edges,[], elevateEdgesOnSelect: true}`

**Reply by bcakmakoglu** [reply]
Can you try passing nodes, edges and elevateEdgesOnSelect as props as well instead of options to `useVueFlow`?

**Reply by DoaaRadwan** [reply]
```
    <VueFlow
      :class="{ dark }"
      :apply-default="false"
      :default-zoom="1"
      :min-zoom="0.2"
      :max-zoom="4"
      :nodes="a"
      :elevate-edges-on-select="true"
      class="basicflow"
      fit-view-on-init
      @dragover="onDragOver"
    >
```
```
} = useVueFlow();
let a = [{ id: "000", label: "Doaa", position: { x: 20, y: 20 } }];
```
still no luck

**Reply by bcakmakoglu** [reply]
Mh... what version are you using?

**Reply by DoaaRadwan** [reply]
1.23.0

**Reply by bcakmakoglu** [reply]
Can you try updating to the latest version and see if that changes anything?

**Reply by bcakmakoglu** [reply]
`1.25.3`

**Reply by DoaaRadwan** [reply]
Thanks, updating and passing `:nodes` `:edges` and `:elevate-edges-on-select` as props instead of options did the trick 

```
onNodesChange((changes) => {
  changes.forEach((change) => {
    if (change.type == "remove") {
      selectedNode = findNode(change.id);
      openDeleteModal();
    } else {
      applyNodeChanges([change]);
    }
  });
});
```
Now the modal on confirm will trigger a function that will use `removeNodes([])`, how would I make `removeNodes([])` take effect ?

**Reply by bcakmakoglu** [reply]
Store the change and apply it, that'd be the easiest way I'd say.

Or store the node id that was supposed to be deleted and then call `removeNodes(nodeId)`

**Reply by bcakmakoglu** [reply]
Something like this

```ts
openDeleteModal({
	onConfirm: () => applyNodeChanges([change])
})
```

**Reply by DoaaRadwan** [reply]
This does work and I have progresses a bit using it however it's really difficult when applying for nodes that has nested nodes inside and you would want the confirmation modal to open for the selected node one and if confirmed you would delete all nodes inside - but since the way onNodesChange works, it will loop through all nodes inside. 

I think I will just disable the delete via backspace for now till I figure a better way to deal with it.

Thank you for your support !

**Reply by DoaaRadwan** [reply]
I think this problem can be solved by applying `:delete-key-code="false"`  and then triggering our logic on a keypress of backspace if there is an event that fires on a that keypress

**Reply by DoaaRadwan** [reply]
I have managed to get it to partially work using this 

```vue
<template>
 <div @keyup.delete="openDeleteModalOnKey" class="dndflow" @drop="onDrop">
    <VueFlow
      :class="{ dark }"
      :default-zoom="1"
      :elevate-edges-on-select="true"
      :delete-key-code="false"
      :min-zoom="0.2"
      :max-zoom="4"
      class="basicflow"
      fit-view-on-init
      @dragover="onDragOver"
    >
    </VueFlow>
 </div>
</template> 

<script setup>
onEdgeClick(({ edge }) => {
  selectedEdge = edge;
});

onNodeClick(({ node }) => {
  selectedNode = node;
});

const openDeleteModalOnKey = () => {
  toggleModalVisibility("deleteModal");
};

</script>
```
it's working as expected for nodes, still trying to get it to work on edges as the keyup-delete event is not firing for edges (at least the ones inside a parent node)
